doc/solver.md

```diff
@@ -0,0 +1,710 @@
+conflict and chooses a different version. However, unlike many search
+algorithms, it also records the root cause of that conflict. This is the
+"conflict-driven clause learning" that lends CDCl its name.
+
+Recording the root causes of conflicts allows Pubgrub to avoid retreading dead
+ends in the search space when the context has changed. This makes the solver
+substantially more efficient than a naïve search algorithm when there are
+consistent causes for each conflict. If no solution exists, clause learning also
+allows Pubgrub to explain to the user the root causes of the conflicts that
+prevented a solution from being found.
+
+# Definitions
+
+## Term
+
+The fundamental unit on which Pubgrub operates is a `Term`, which represents a
+statement about a package that may be true or false for a given selection of
+package versions. For example, `foo ^1.0.0` is a term that's true if `foo 1.2.3`
+is selected and false if `foo 2.3.4` is selected. Conversely, `not foo ^1.0.0`
+is false if `foo 1.2.3` is selected and true if `foo 2.3.4` is selected.
+
+We say that a term `t` is "satisfied" by a set of terms `S` if `t` must be true
+whenever every term in `S` is true. Conversely, `t` is "contradicted" by `S` if
+it must be false whenever every term in `S` is true. If neither of these is
+true, we say that `S` is "inconclusive" for `t`. As a shorthand, we say that a
+term `v` satisfies or contradicts `t` if `{v}` satisfies or contradicts it. For
+example:
+
+* `{foo >=1.0.0, foo <2.0.0}` satisfies `foo ^2.0.0`,
+* `foo ^1.5.0` contradicts `not foo ^1.0.0`,
+* and `foo ^1.0.0` is inconclusive for `foo ^1.5.0`.
+
+Terms can be viewed as denoting sets of allowed versions, with negative terms
+denoting the complement of the corresponding positive term. Set relations and
+operations can be defined accordingly. For example:
+
+* `foo ^1.0.0 ∪ foo ^2.0.0` is `foo >=1.0.0 <3.0.0`.
+* `foo >=1.0.0 ∩ not foo >=2.0.0` is `foo ^1.0.0`.
+* `foo ^1.0.0 \ foo ^1.5.0` is `foo >=1.0.0 <1.5.0`.
+
+This turns out to be useful for computing satisfaction and contradiction. Given
+a term `t` and a set of terms `S`, and letting `∩S` denote the intersection of
+all terms in `S`, we have the following identities:
+
+* `S` satisfies `t` if and only if `∩S ⊆ t`.
+* `S` contradicts `t` if and only if `∩S` is disjoint with `t`.
+
+## Incompatibility
+
+An incompatibility is a set of terms that cannot all be true—that is, at least
+one term in every incompatibility must be false for a given set of package
+versions in order for it to be considered a valid solution. Incompatibilities
+are *context-independent*, meaning that their terms are mutually incompatible
+regardless of which versions are selected at any given point in time.
+
+There are two sources of incompatibilities. They may represent facts about
+packages—for example, "`foo ^1.0.0` depends on `bar ^2.0.0`" is represented as
+the incompatibility `{not foo ^1.0.0, bar ^2.0.0}`, while "`foo <1.3.0` has an
```

Comment:
Looking at the [first example](https://github.com/dart-lang/pub/blob/e891bfbfaaddf07b7906abeb424cda8c8efd8579/doc/solver.md#no-conflicts) the `not` seems to be on the other side. Shouldn't it be `{foo ^1.0.0, not bar ^2.0.0}` ?